mmv1/third_party/terraform/services/container/resource_container_node_pool.go.tmpl

```diff
@@ -1728,13 +1672,24 @@ func containerNodePoolAwaitRestingState(config *transport_tpg.Config, name, proj
 	return state, err
 }
 
-// Retries an operation while the canonical error code is FAILED_PRECONDTION
-// or RESOURCE_EXHAUSTED which indicates there is an incompatible operation
-// already running on the cluster or there are the number of allowed
-// concurrent operations running on the cluster. These errors can be safely
-// retried until the incompatible operation completes, and the newly
-// requested operation can begin.
-func retryWhileIncompatibleOperation(timeout time.Duration, lockKey string, f func() error) error {
+// Retries an operation while the canonical error code is FAILED_PRECONDTION or RESOURCE_EXHAUSTED which indicates
+// there is an incompatible operation already running on the cluster or there are the number of allowed concurrent
+// operations running on the cluster. These errors can be safely retried until the incompatible operation completes,
+// and the newly requested operation can begin.
+// The cluster lock is held during createOpFunc to make opeation creations sequencial, and cluster read lock is held
+// during waitOpFunc to allow concurrency.
+func retryWhileIncompatibleOperation(timeout time.Duration, lockKey string, clusterLockKey string, createOpFunc func() (*container.Operation, error), waitOpFunc func(*container.Operation) error) error {
+	f := func() error {
+		transport_tpg.MutexStore.Lock(clusterLockKey)
+		op, err := createOpFunc()
+		transport_tpg.MutexStore.Unlock(clusterLockKey)
+		if err != nil {
+			return err
+		}
+		transport_tpg.MutexStore.RLock(clusterLockKey)
+		defer transport_tpg.MutexStore.RUnlock(clusterLockKey)
+		return waitOpFunc(op)
+	}
```

Comment:
So I don't believe this to actually be changing anything after going through the behavior loop in my head. `LockedCall(createLockKey, ...)` holds the lock for the entire operation (both creation and waiting). This means the additional locking/unlocking with `waitLockKey` inside the operation serves no purpose, as the outer lock already prevents concurrent execution.

Also, I don't believe concurrent updates are supported in the provider. all operations right now are concurrent

Reply:
So the goal is that we don't want multiple nodepools within the same cluster to create operations at exactly the same time. Currently, users may trigger creation of multiple (a lot of) nodepools for a single cluster using terraform and they will be created at the exact same time, overwhelming the GKE pipeline. The nodepool lock is fine, because we want sequential creation at cluster level.
Also happy to chat more on GVC.

Reply:
I believe I understand it a bit better after drawing a diagram. I worry a bit about the readability of this code given that need. I'm okay approving this as is, but could someone from your team also review because the maintenance will fall more broadly there.

```mermaid
sequenceDiagram
    participant Caller
    participant Retry
    participant LockedCall
    participant MutexStore
    participant CreateOp as createOpFunc
    participant WaitOp as waitOpFunc

    Caller->>Retry: retryWhileIncompatibleOperation(timeout, npLockKey, clusterLockKey, ...)
    
    activate Retry
    Note over Retry: Start retry loop with timeout
    
    Retry->>LockedCall: LockedCall(npLockKey, f)
    activate LockedCall
    
    LockedCall->>MutexStore: Lock(npLockKey)
    Note over MutexStore: Write lock on nodepool
    
    LockedCall->>MutexStore: Lock(clusterLockKey)
    Note over MutexStore: Write lock on cluster
    
    LockedCall->>CreateOp: createOpFunc()
    activate CreateOp
    CreateOp-->>LockedCall: Return operation
    deactivate CreateOp
    
    LockedCall->>MutexStore: Unlock(clusterLockKey)
    Note over MutexStore: Release cluster write lock
    
    LockedCall->>MutexStore: RLock(clusterLockKey)
    Note over MutexStore: Read lock on cluster
    
    LockedCall->>WaitOp: waitOpFunc(op)
    activate WaitOp
    WaitOp-->>LockedCall: Return result
    deactivate WaitOp
    
    LockedCall->>MutexStore: RUnlock(clusterLockKey)
    Note over MutexStore: Release cluster read lock
    
    LockedCall->>MutexStore: Unlock(npLockKey)
    Note over MutexStore: Release nodepool lock (via defer)
    
    alt Operation succeeded
        LockedCall-->>Retry: Return nil
        Retry-->>Caller: Return nil
    else IsFailedPreconditionError or IsQuotaError
        LockedCall-->>Retry: Return error
        Note over Retry: Retry operation
        Retry->>LockedCall: LockedCall(npLockKey, f)
    else Other error
        LockedCall-->>Retry: Return error
        Retry-->>Caller: Return error immediately
    end
    
    deactivate LockedCall
    deactivate Retry
```

Reply:
Alex from the team has reviewed the PR.

Reply:
Perfect. Are there any other changes you wish to make? If not I can approve. I want to save space for that if so.

Reply:
Now it's ready : )